### PR TITLE
GDScript: Add `@warning_ignore_start` and `@warning_ignore_restore` annotations

### LINKED ISSUE #5906: Ignore warnings for multiple lines in GDScript
### Describe the project you are working on

any GDScript code

### Describe the problem or limitation you are having in your project

I want to be able to ignore a GDScript warning (or potentially multiple) for multiple lines of code and enable the warning again after these lines.

currently it is only possible to ignore a warning for one line or for the entire file

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

add a new warning directive to GDScript similar to `# warning-ignore-all: X`, but for whole blocks of code

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

```gdscript
... some warning free code ...

# warning-ignore-start: X, Y
... some code block which triggers warnings X and Y ...
... more lines ...
... more lines ...
...
# warning-ignore-end: X, Y

... more warning free code ...
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

every line which triggers warning X needs to be prepended by `# warning-ignore: X`, making the code and real comments hard to read

### Is there a reason why this should be core and not an add-on in the asset library?

core functionality of GDScript

### LINKED ISSUE #93202: Signals from signal bus are marked as "not used in the class"
### Tested versions

v4.3.beta1.official [a4f2ea91a]
(not happening in 4.2.2)

### System information

Godot v4.3.beta1 - macOS 14.5.0 - Vulkan (Forward+) - integrated Apple M1 - Apple M1 (8 Threads)

### Issue description

Signal bus is a pattern that Godot practically makes obligatory. It is often (from what I saw of other people's code) implemented as one of autoloads, usually as a script containing only and dedicated exclusively to signals.

Since 4.3, every signal in such a script is producing a warning.
image
This can get quite annoying when there's more than a few of these.

I understand that I can just add an "ignore warning" for each signal and I understand that the way this is detected is actually more logical than what we had before (since indeed what the warning is reporting is technically true), but I think it makes no sense to require adding an "ignore warning" every time I create a new signal in this type of autoload.

### Steps to reproduce

Make a signal bus, launch a project, read the warnings.

### Minimal reproduction project (MRP)

Any.

### PR DESCRIPTION
**Pull request by dalexeev**
* Closes godotengine/godot-proposals#5405.
* Closes godotengine/godot-proposals#5906.
* Resolves #93202.
* **Also removes a controversial undocumented [feature](https://github.com/godotengine/godot/pull/83037#discussion_r1477302018). Instead, you can wrap the function in the annotations.**

### CODE REVIEW

File: modules/gdscript/gdscript_parser.h [hunk 1]
```diff
@@ -1283,6 +1283,8 @@ class GDScriptParser {
 	bool is_ignoring_warnings = false;
 	List<GDScriptWarning> warnings;
 	HashSet<GDScriptWarning::Code> ignored_warnings;
+	HashSet<int> script_ignored_warning_lines[GDScriptWarning::WARNING_MAX];
+	HashMap<GDScriptWarning::Code, int> script_ignored_warning_start_lines;
 	HashSet<int> unsafe_lines;
```
Comment (dalexeev) [inline modules/gdscript/gdscript_parser.h:1288]: I'm not sure if I used the correct structures.

File: modules/gdscript/gdscript_parser.cpp [hunk 2]
```diff
@@ -4055,6 +4090,46 @@ bool GDScriptParser::warning_annotations(const AnnotationNode *p_annotation, Nod
...
+			}
+		}
+	}
+
+	return !has_error;
+
+#else // ! DEBUG_ENABLED
+	// Only available in debug builds.
+	return true;
+#endif // DEBUG_ENABLED
```
Comment (adamscott) [inline modules/gdscript/gdscript_parser.cpp:4130]: I would switch the test to make the `!DEBUG_ENABLED` macro return early.

File: modules/gdscript/gdscript_parser.h (hunk 1 above)
Comment (adamscott) [reply modules/gdscript/gdscript_parser.h:1292]: https://github.com/godotengine/godot/blob/c80a2b4fe99dcd0bba6fc24ed2748b1474b24448/core/templates/hash_map.h#L41-L45

Comment (adamscott) [reply modules/gdscript/gdscript_parser.h:1292]: From what I read, Robin Hood swaps the entries, so if you do a `for each` on the hashmap, it's not guaranteed that you'll get each entry. If you're only getting the value that you want, it seems ok.

File: modules/gdscript/tests/scripts/parser/features/warning_ignore_regions.gd
[the remark this answers is not in the document]
Comment (dalexeev) [reply modules/gdscript/tests/scripts/parser/features/warning_ignore_regions.gd:1]: I have added more tests.

File: modules/gdscript/doc_classes/@GDScript.xml [hunk 3]
```diff
@@ -742,6 +743,33 @@
...
+			<description>
+				Restores ignoring the listed warning types after [annotation @warning_ignore_start] to the state specified in the Project Settings. This annotation is optional, you are not required to use it if you want to ignore some warning type until the end of the file.
+				[b]Note:[/b] Unlike most other annotations, arguments of the [annotation @warning_ignore_restore] annotation must be string literals (constant expressions are not supported).
+			</description>
+		</annotation>
+		<annotation name="@warning_ignore_start" qualifiers="vararg">
+			<return type="void" />
+			<param index="0" name="warning" type="String" />
+			<description>
+				Enables ignoring the listed warning types until the end of the file or until the parser encounters [annotation @warning_ignore_restore] with this warning type.
```
Comment (Mickeon) [inline modules/gdscript/doc_classes/@GDScript.xml:761]: This reads like a double negation. From the outside user's perspective the "parser encountering something" is a bit unusual to say, but I have no suggestions at the moment.
```suggestion
				Ignores the listed warning types until the end of the file or until the parser encounters [annotation @warning_ignore_restore] with the given warning types.
```

File: modules/gdscript/doc_classes/@GDScript.xml [hunk 4]
```diff
@@ -742,6 +743,33 @@
...
 				    print("unreachable")
 				[/codeblock]
+				See also [annotation @warning_ignore_start] and [annotation @warning_ignore_restore].
+			</description>
+		</annotation>
+		<annotation name="@warning_ignore_restore" qualifiers="vararg">
+			<return type="void" />
+			<param index="0" name="warning" type="String" />
+			<description>
+				Restores ignoring the listed warning types after [annotation @warning_ignore_start] to the state specified in the Project Settings. This annotation is optional, you are not required to use it if you want to ignore some warning type until the end of the file.
```
Comment (Mickeon) [inline modules/gdscript/doc_classes/@GDScript.xml:753]:
```suggestion
				Stops ignoring the listed warning types after [annotation @warning_ignore_start] to the state specified in the Project Settings. This annotation can be omitted to ignore the warning types until the end of the file.
```

**Review by akien-mga — APPROVED** [review]
Looks great to me!

File: modules/gdscript/gdscript_parser.cpp [hunk 5]
```diff
@@ -4831,6 +4875,45 @@ bool GDScriptParser::warning_annotations(AnnotationNode *p_annotation, Node *p_t
 #endif // DEBUG_ENABLED
 }
 
+bool GDScriptParser::warning_ignore_region_annotations(AnnotationNode *p_annotation, Node *p_target, ClassNode *p_class) {
+#ifndef DEBUG_ENABLED
+	// Only available in debug builds.
```
Comment (AThousandShips) [inline modules/gdscript/gdscript_parser.cpp:4880]: I'd flip this condition for readability, i.e.:
```cpp
#ifdef DEBUG_ENABLED
...
#else
...
#endif
```

**Review by AThousandShips — APPROVED** [review]

File: modules/gdscript/gdscript_parser.cpp (hunk 5 above)
Comment (dalexeev) [reply modules/gdscript/gdscript_parser.cpp:4880]: See https://github.com/godotengine/godot/pull/76020#discussion_r1190040033. Funny, we've come full circle (this was changed in another PR, then this one was rebased).

**Comment by Repiteo** [comment]
Thanks!

**Pull request merged**